auth/auth0_manager.go

```diff
@@ -36,16 +41,32 @@ const (
 	validAuthExpiry = 1 * time.Hour
 )
 
-func NewAuth0Manager(domain, clientID, connection string) *Auth0Manager {
+func NewAuth0Manager(domain, clientID, connection string, throttleMin int) *Auth0Manager {
 	return &Auth0Manager{
-		clientID:   clientID,
-		connection: connection,
-		client:     rclient.NewRestClient(domain),
-		cache:      cache.New(),
+		clientID:           clientID,
+		connection:         connection,
+		client:             rclient.NewRestClient(domain),
+		cache:              cache.New(),
+		throttle:           time.Tick(time.Millisecond * time.Duration(throttleMin)),
+		throttleMin:        throttleMin,
+		throttleMultiplier: throttleMin,
+		lastThrottled:      time.Now(),
 	}
 }
 
 func (a *Auth0Manager) Authenticate(username, password string) (bool, error) {
+	if a.throttleMultiplier > a.throttleMin && time.Since(a.lastThrottled) > 1*time.Minute {
```

Comment:
Right - this logic is decreasing the backoff.

The increase comes when a 429 is returned from making the authentication, and the throttle is waited for between the token auth and the auth0 auth; that way, requests using a valid token aren't throttled.

This bit of logic I put at the very start because I wanted the checks for decreasing the delay to happen on every request. It avoids a scenario like the following:

```
1. On a Monday, someone generates a swarm of bad auths and ratchets the throttle up to like 32 min, then stops.
2. All throughout the rest of the week, nobody makes a request without a valid token, so the decrease-throttle logic is never run
3. Next week, someone tries to authenticate using Auth0 for the first time in a week, and they still have to sit through 32 (or 16, depending on where this logic sits) min for last week's throttle
```